### PR TITLE
Sleep the coreaudio driver in-editor after 15s of silence

### LINKED ISSUE #28039: Godot prevents automatic sleeping due to the audio system
___

***Bugsquad note:** This issue has been confirmed several times already. No need to confirm it further.*

___

**Godot version:** 3.1

**OS/device including version:** Mac OSX 10.14

**Issue description:** 
The godot editor is preventing Mac's from automatically sleeping, according to the activity monitor.

Screen Shot 2019-04-14 at 7 45 14 PM

**Steps to reproduce:** Open any godot project.

### PR DESCRIPTION
**Pull request by ellenhp**
Fixes #28039
Should help mitigate #38154 but might not solve the root cause

This PR allows the coreaudio driver to enter a sleep state when the AudioServer requests it to. Mutexes have been added to all audio players and the video player to prevent a play() call acquiring a playback "lock" from racing with a _mix_audio call releasing it. I'm not a C++ expert so maybe there's some fancy RAII way to do this but I figured I'd start simple and avoid overengineering.

How to test:

1. Build & run the editor with these changes on a mac
2. Create each kind of AudioStreamPlayer. Play the players one at a time and watch Godot's "Preventing Sleep" status in Activity Monitor flip from "No" to "Yes" within a few seconds of starting the player, then flip back to "No" about 15 seconds after the audio is finished playing.

I have been unable to test video players. They don't seem to work in latest master as far as I can tell.

### CODE REVIEW

File: drivers/coreaudio/audio_driver_coreaudio.cpp
Comment (naithar) [inline drivers/coreaudio/audio_driver_coreaudio.cpp:246]: I think checking `active` first like this `if (active) { return; }` to perform early exit and then checking `sleeping` should make it a bit more readable.

File: servers/audio_server.cpp
Comment (naithar) [inline servers/audio_server.cpp:1052]: If this change should only work for editor, maybe `#ifdef TOOLS_ENABLED` should also be added, so it's not included in templates builds?

**Comment by naithar** [comment]
I'll try testing it once I'm free, but so far changes look okay.

File: servers/audio_server.cpp
Comment (ellenhp) [reply servers/audio_server.cpp:1052]: The main reason I see to do that is export binary size and potentially (slightly) reducing lock contention, which makes me wonder if I should do it elsewhere too. What do you think?

Comment (RandomShaper) [reply servers/audio_server.cpp:1052]: I haven't yet checked the whole PR, but I can already say if you want something to happen only in the editor the current check is better, because it is `false` both on tools builds when running a project (vs. running the editor) and on non-tools builds.
Furthermore, in non-tools builds, given the check is known to fail at compile time, I'm quite certain that compilers can optimize out the body.

File: drivers/coreaudio/audio_driver_coreaudio.cpp
Comment (ellenhp) [reply drivers/coreaudio/audio_driver_coreaudio.cpp]: This complicated the diffs a bit for anyone reviewing, but I agree it's more readable when viewed on its own. Fixed.

**Comment by RandomShaper** [comment]
I'm thinking this could be simpler. Let me explain.

We have the `audio/channel_disable_time` project setting, that defaults to 2 seconds, but let's assume the editor could override it internally to 15.

Now, `AudioServer::_mix_step()` goes through all the buses and items in it and is able to flag as inactive those that meet the condition for inactivation (i.e., too much time without playing):
https://github.com/godotengine/godot/blob/2b95372ad1470e374112c5424848eea501465b86/servers/audio_server.cpp#L446

Because of that, and more importantly, it has a notion about if any channel is playing or none at all. This check can be used to contribute to a local boolean that tells that once the loop has ended:
https://github.com/godotengine/godot/blob/2b95372ad1470e374112c5424848eea501465b86/servers/audio_server.cpp#L403

My point is: couldn't that be used to make the driver sleep and wake up in a simpler that than all the added code? (The added API to the drivers would still be needed and leveraged.)

**Comment by ellenhp** [comment]
That's similar to the way I wanted to do this fix originally, but that code to check if a bus is active in `_mix_step` won't be called if the audio driver is sleeping.

The CoreAudio driver sets up its own `output_callback` function to be called by CoreAudio, I believe. That callback calls `audio_server_process` which calls `_driver_process` which calls `_mix_step`. Without that callback actively being called by CoreAudio, the only way to restart the driver would be from outside the audio thread.

**Comment by RandomShaper** [comment]
Oh, I wasn't aware of that. Makes sense.

File: scene/2d/audio_stream_player_2d.h
Comment (RandomShaper) [inline scene/2d/audio_stream_player_2d.h:92]: The several `playback_lock_mutex` are for making the various player nodes thread-safe in regards to the so-called playback lock, right? Nodes in general are not required to be thread-safe and users are expected to do their own locking. Therefore, maybe those mutexes can go away.

Not super important, though. I'd like to read more opinions on this.

File: servers/audio_server.cpp
Comment (RandomShaper) [inline servers/audio_server.cpp:927]: To manage `playing_sources_count` safely you can just use it as an atomic and avoid the mutex. To do it Godot-style (by the way, #45617 will introduce a new style):
```
uint32_t playing_sources_count;

// [...]

void AudioServer::notify_source_playing() {
    if (atomic_increment(&playing_sources_count) == 1) {
        AudioDriver::get_singleton()->set_sleep_state(false);
    }
}

void AudioServer::notify_source_stopped_playing() {
    if (atomic_decrement(&playing_sources_count) == 0) {
        last_playback_time_msec = OS::get_singleton()->get_ticks_msec();
    }
}
```

File: scene/2d/audio_stream_player_2d.h
Comment (ellenhp) [reply scene/2d/audio_stream_player_2d.h:92]: I'm a bit more worried about the `play()` call racing with the audio thread's `_mix_audio()` calls, but I guess that two `play()` calls could also race, or potentially a `play()` and a an exit tree notification even.

This code difficult to reason about, and it's not entirely clear how high the risk of a race condition actually is, but it is clear to me that if a race does occur, the audio driver's count of sources that are currently playing could be thrown off by one, which would cause it to never sleep or (worse) not wake up despite there being a source that wants to play audio.

By setting the bool and calling the `notify_source_playing`/`notify_source_stopped_playing` atomically I'm avoiding that risk. I'm also very interested to hear other thoughts though, since my C++ is a bit rusty.

**Comment by fab918** [comment]
Is this improvement can be available for release build? (useful if you make a software)

**Comment by ellenhp** [comment]
@fab918 I'm hoping this fix will be cherry-picked for 3.2 so those of us who prefer stable builds (like me 😄) can benefit. I don't know if it will make it into 3.2.4 though. It may have to wait until the release after that. It really depends on how many more release candidates there are and what the project maintainers are comfortable cherry picking between release candidates. It would be nice though to get into 3.2.4 but I need to remind myself to be patient sometimes 👍 

I'll respond to the most recent round of review comments tomorrow morning UTC-8.

File: servers/audio_server.cpp
Comment (naithar) [reply servers/audio_server.cpp:1052]: Well, some places seem to use only `Engine::get_singleton()->is_editor_hint()` while others use both `Engine::get_singleton()->is_editor_hint()` and `TOOLS_ENABLED` checks. Maybe I'm missing something, but if using only `Engine::get_singleton()->is_editor_hint()` is preferred it's good to know :)

**Comment by naithar** [comment]
Tested on M1 MacBook, but I don't think result would be different for Intel based macs. 

When running Project Editor everything seems to work exactly as described - after some time `coreaudiod` returns to 0% CPU and Godot is no longer preventing mac from going to sleep mode.
But Project Manager doesn't seems to be affected, so it's still preventing sleep and taking CPU time through `coreaudiod`. I'm not sure it's anything serious, but if it's possible to apply this fix to all editor windows I think it should be applied.

**Comment by Calinou** [comment]
> But Project Manager doesn't seems to be affected, so it's still preventing sleep and taking CPU time through coreaudiod. I'm not sure it's anything serious, but if it's possible to apply this fix to all editor windows I think it should be applied.

https://github.com/godotengine/godot/pull/38208 would fix this :slightly_smiling_face:

**Comment by ellenhp** [comment]
> But Project Manager doesn't seems to be affected, so it's still preventing sleep and taking CPU time through coreaudiod. I'm not sure it's anything serious, but if it's possible to apply this fix to all editor windows I think it should be applied.

I haven't confirmed this, but I suspect the issue here is that `is_editor_hint()` returns false in the project manager.

**Comment by naithar** [comment]
> > But Project Manager doesn't seems to be affected, so it's still preventing sleep and taking CPU time through coreaudiod. I'm not sure it's anything serious, but if it's possible to apply this fix to all editor windows I think it should be applied.
> 
> I haven't confirmed this, but I suspect the issue here is that `is_editor_hint()` returns false in the project manager.

Yeah, seems like it. But as @Calinou mentioned #38208 should fix it, so that should be okay. 
I'll retest again after you make other changes, but everything looks good so far.

File: servers/audio_server.cpp
Comment (ellenhp) [reply servers/audio_server.cpp]: Done!

**Review by RandomShaper — APPROVED** [review]
I think others need still to give a final review. For me, as far as I'm concerned, this is good enough the way it is now.

Also, @ellenhp, thanks for your patience and nicely accepting the feedback!

**Comment by ellenhp** [comment]
No problem! Code review is like, really important for the health of the project and as a user of Godot I care a lot about the health of the project! 😄 I'll work on a sister PR for 3.2. There will be some small changes to mutexes and I think some of the files may have been renamed, but it should largely be the same. I'll leave it in a draft state until this one is merged though.

**Comment by naithar** [comment]
Everything seems to work correctly, so LGTM.

**Comment by akien-mga** [comment]
Changes look good to me overall, though it would be good to have a quick review from @reduz as there are changes to AudioServer.

Some comments on the form:

- It would be good to squash the three commits into one, as the last two comments amend/refactor what the first commit did. We usually prefer to have the "final" state of a PR reflected in the commits we merge, so that potential bisecting or post-merge code review doesn't need to jump through several iterations of code design. See [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase) for instructions on how to rebase/squash if needed.

- Your commits seem not to be attributed to your GitHub account. Most likely the email used (see header in https://patch-diff.githubusercontent.com/raw/godotengine/godot/pull/45948.patch) is not linked to your @ellenhp account. It's not a problem for Git, but if you want to claim ownership of those commits, you can add this email as secondary email in your GitHub settings.

File: servers/audio_server.cpp
Comment (akien-mga) [reply servers/audio_server.cpp]: There's no clear policy, but yes we sometimes put both checks in place to ensure that the editor-specific code (often expensive debug code) is compiled out in the runtime. If @RandomShaper is correct that compiler would do it just fine with `is_editor_hint()` only, then we shouldn't need to do that indeed.

**Comment by akien-mga** [comment]
Regarding the use of `Engine::get_singleton()->is_editor_hint()` and it not applying to the project manager, another option would be to use a project setting to control this behavior, and turn it on in `main.cpp` for both the `editor` and the `project_manager`.

This would make it possible to use this in Godot projects too, e.g. when making games without sound, or non-game apps.

**Comment by bruvzg** [comment]
> I haven't confirmed this, but I suspect the issue here is that `is_editor_hint()` returns false in the project manager.

It is `false` in project manager, use `Main::is_project_manager()` to check for PM (include `main\main.h`).

File: servers/audio_server.cpp
Comment (naithar) [reply servers/audio_server.cpp]: If we want to have access to this feature enabled for any other engine mode in project settings, then I guess `TOOLS_ENABLED` isn't needed too.

**Comment by reduz** [comment]
Some thoughts on this PR..

* I don't like the added complexity in AudioStreams
* In many situations, you will use streams continually toggled on (like voip, microphone, etc).

If the problem is actually not being able to sleep on macs when the editor is turned on, I would probably prefer to not stop the audio thread but stop Coreaudio after some time with silence, and then keep a thread alive checking if there is audio data in order to re-enable coreaudio. This of course would be used in editor or lowpcu mode only.

**Comment by reduz** [comment]
I also wonder if this could probably be done in AudioServer directly, so drivers also dont really know whats going on either. When low cpu mode is enabled, turn off the audio driver after X time of silence and start a thread that scans for any audio going on, then re-initializes audio.

**Comment by ellenhp** [comment]
Could you expand on what you mean by stopping CoreAudio without stopping the audio thread? If that's possible I think it would be the way to go, but I'm not sure how I'd go about that. It looks to me like stopping CoreAudio stops the audio thread, but I could be wrong.

**Comment by ellenhp** [comment]
This PR will get way simpler if I manage to implement https://github.com/godotengine/godot-proposals/issues/2299 so I think I'm going to close it. The downside is that the implementation of that proposal will almost certainly not be backported, so fixing this bug will probably be something that only happens in 4.0.

**Comment by Calinou** [comment]
It looks like this approach would be useful to explore again, with https://github.com/godotengine/godot/issues/28039 also occurring on Windows.

**Comment by ellenhp** [comment]
Interesting. This is a much easier fix now after the AudioServer rewrite because we don't have to add code in each player node anymore. The audio server has all the information it needs to toggle the driver on and off.

**Pull request closed without merging**
